store/streaming/file/service.go

```diff
@@ -20,262 +22,150 @@ var _ baseapp.StreamingService = &StreamingService{}
 
 // StreamingService is a concrete implementation of StreamingService that writes state changes out to files
 type StreamingService struct {
```

Comment:
The (exported) methods on this type aren't safe for concurrent use by multiple goroutines. Is it guaranteed that implementations of StreamingService will never be accessed concurrently? If so, is that guarantee documented somewhere?

Reply:
Yeah, I believe there's no concurrency here because it's only called by consensus state machine.

Reply:
[The ADR proposes implementations of](https://github.com/cosmos/cosmos-sdk/blob/f1204486904285d0d862ada3dc34318a4586ee54/docs/architecture/adr-038-state-listening.md#exposing-the-data) BeginBlock, EndBlock, DeliverTx, etc. which could result in concurrent use of the same listener by multiple goroutines if HaltAppOnDeliveryError is false. (It could also happen when HaltAppOnDeliverError is true, if for any reason those ABCI methods weren't serialized by the application.)

Reply:
- As far as the concern of cosmos-sdk, the abci methods are executed serialized, I believe many code rely on that assumption.
- I've left my comment there as well to suggest leaving the decision of go-routine to listener implementation: https://github.com/cosmos/cosmos-sdk/pull/13473#discussion_r1022204663

Reply:
Specifically, "fire-and-forget" goroutines spawned by BaseApp methods like BeginBlock, DeliverTx, etc. as so:

```go
func (app *BaseApp) BeginBlock(req abci.RequestBeginBlock) (res abci.ResponseBeginBlock) {
	...
	for _, streamingListener := range app.abciListeners {
		...
		// fire and forget semantics
		go func() {
			if err := streamingListener.ListenBeginBlock(app.deliverState.ctx, req, res); err != nil {
				app.logger.Error("BeginBlock listening hook failed", "height", req.Header.Height, "err", err)
			}
		}()
```

will not be finished when the outer BaseApp method returns. That means whatever caller invoked e.g. BeginBlock could proceed to invoke e.g. EndBlock, which would spawn an equivalent goroutine, accessing the same streamingListener:

```go
func (app *BaseApp) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBlock) {
	...
	for _, streamingListener := range app.abciListeners {
			// fire and forget semantics
			go func() {
				if err := streamingListener.ListenEndBlock(app.deliverState.ctx, req, res); err != nil {
					app.logger.Error("EndBlock listening hook failed", "height", req.Height, "err", err)
				}
			}()
```

The "fire-and-forget" goroutines spawned by (in this example) BeginBlock and EndBlock would then concurrently access the same streamingListener. As that type is not synchronized, this would result in a memory model violation.

Reply:
You are still talking about #13473, right? because in this PR, there's no go-routine here.

Reply:
I'm quoting [docs/architecture/adr-038-state-listening.md as it exists in this PR](https://github.com/yihuang/cosmos-sdk/blob/fix-listener-main3/docs/architecture/adr-038-state-listening.md) which uses goroutines in the methods I referred to. Is that ADR markdown file not correct?

Reply:
> I see, but those don't match the implementation, it's a mess, I prefer to update the ADR for this, so the implementation will be simpler.

Okay, happy to re-review when the ADR is updated to match reality.

Reply:
> Anyway, in general, I hope the builtin streaming services is simple, minimal, low overhead, reliable by default, like the file streamer, and we can always https://github.com/cosmos/cosmos-sdk/issues/13652 externally for async complex cases, WDYT?

"Reliable delivery" is unfortunately not well-defined. I guess most distsys literature defines "reliable delivery" to include, among other things, that the information was received by the recipient(s) in a way which is verifiable by the sender(s). Writing a file successfully asserts (a restricted form of) information availability, but the filesystem doesn't provide any way for file _readers_ to convey their consumption of that information to file _writers_ without some additional protocol. 

I agree that the file streamer provides a simple, minimal, [relatively] low-overhead way of emitting state updates. But it doesn't, and can't, provide _reliable_ delivery. It can only provide best-effort delivery.

Reply:
By "reliable delivery", I always mean "at least once" message delivery in my head, WDYT if we change it to "at least once" msg delivery guarantee within reasonable time period?

Reply:
"At least once" and "at most once" and "exactly once" are semantics defined in terms of message receivers, not message producers. That is, "at least once" means the message must be successfully delivered at least once to its recipients, and must not be lost.

Systems like Kafka can provide this guarantee because they can do things like hold on to a message until it is successfully consumed. Filesystems cannot provide this kind of guarantee because they do not, by themselves, provide any guarantees about file _reads_. 

Filesystems can guarantee that a write will be durable and available for future reads — to some extent, and as long as you explicitly fsync, and etc. But filesystems cannot guarantee that a write will actually be read by anyone, in any amount of time. So a filesystem cannot by itself meet "at least once" or "at most once" or "exactly once" delivery requirements. It simply doesn't have enough information.

(Now, you can define a protocol _on top_ of a filesystem that expresses reads as deletes, or moves, or etc., and if you have a sound multi-user filesystem implementation, you can get at-most-once or at-least-once or whatever other delivery semantics out of that protocol. But that's not what we're talking about here.)

Reply:
> Filesystems can guarantee that a write will be durable and available for future reads — to some extent, and as long as you explicitly fsync, and etc.

we do provide the optional fsync call.

> But filesystems cannot guarantee that a write will actually be read by anyone, in any amount of time.

that's why I added the clause "before the historical data is pruned away", the pruning logic could be a separate issue, and can be done externally through filesystem apis.
I'm open to better wordings to the "reliable delivery" with clauses, any suggestions?

Reply:
> that's why I added the clause "before the historical data is pruned away", the pruning logic could be a separate issue, and can be done externally through filesystem apis by user according to the needs
> . . . 
> I'm open to better wordings to the "reliably delivery" with clauses, any suggestions?

Sure — "best-effort"! That's totally fine. You can do a lot of useful stuff with best-effort event delivery semantics. Consumers who want reliable and deterministic information from a node can already satisfy their needs by polling the existing RPC endpoints.

Reply:
> > that's why I added the clause "before the historical data is pruned away", the pruning logic could be a separate issue, and can be done externally through filesystem apis by user according to the needs
> 
> > . . . 
> 
> > I'm open to better wordings to the "reliably delivery" with clauses, any suggestions?
> 
> 
> 
> Sure — "best-effort"! That's totally fine. You can do a lot of useful stuff with best-effort event delivery semantics. Consumers who want reliable and deterministic information from a node can already satisfy their needs by polling the existing RPC endpoints.

They can't consume state changes with existing rpc endpoints, and with fsync and not pruning, we do match the "at least once" reliable delivery definition here, right?

Reply:
> with fsync and not pruning, we do match the "at least once" reliable delivery definition here, right? 

No. Unfortunately!

"At least once" says that every intended recipient of a given message will get that message, and that there is no chance that it will be lost. Writing a file to disk, even with fsync, and even if you disable pruning, does not provide that guarantee. A filesystem can ensure properties of the write operation, but it doesn't ensure any properties of future reads.

Reply:
> Someone can always come along and rm -rf the file, or the directory, or mv it somewhere else, or any number of things, and the filesystem will not understand that as a problem.

This is beyond the reliable definition here, these kind of bad operations can happen to any system, one can always delete arbitrary db files of kafka or postgresql, or simply kill the process, and the system will not understand that as a problem.

Reply:
> They can't consume state changes with existing rpc endpoints,

If you mean state changes that occur from committed blocks, then this doesn't matter. Blocks are committed atomically, so there's no difference between a single event containing an ordered list of transactions and the ultimate result on app state, compared to breaking that single event up into individual transactions and their delta updates to app state. Say a block is `t1 t2 t3 t4` — there's no way for a listener to do anything between `t2` and `t3`, anything they do will happen after `t4`. So in this case, there is no difference between event streams and RPC queries.

If you mean state changes that occur _before_ committed blocks, like as part of gossiped mempool txs? Gossip is unreliable, so each node has its own unique mempool txs, and a nondeterminnistic tx order. Each node's event stream will be different. There isn't a deterministic, authoritative source of truth for these events. So in this case, event streams are "best effort" by definition!

AFAIU there's nothing in between these two things. So it's OK that there's no way to offer reliable event delivery, because there's no situation where reliable events actually make a difference! 😉

Reply:
> This is beyond the reliability definition here, these kind of bad operations can happen to any system, one can always delete arbitrary db files of kafka or postgresql, or simply kill the process, and the system will not understand that as a problem.

Solar flares can fuck up any system, too, sure. The difference is that e.g. Kafka understands, and tries to enforce, delivery guarantees, whereas filesystems do not.

Reply:
I mean the state changes produced by the state listeners, which is not available in existing RPC interfaces, but anyway, that's irrelevant to our discussion here. actually I'm not sure what we are discussing here. = ="
Let me try to be precise again, with `stopNodeAtErr=true` and `fsync=true`, file streamer output is reliable for the consumers, in the sense that: we won't lose block data in the middle of the stream, in face of node restart or system crash, do you agree with this statement?

Reply:
> with stopNodeAtErr=true and fsync=true, file streamer output is reliable for the consumers, in the sense that: we won't lose block data in the middle of the stream, no matter node restart or system crash, do you agree with this statement? (ofc we need to ignore the mis-operations)

No. This configuration provides reliable (more specifically: durable) write guarantees for _producers_ of events, but doesn't provide any guarantees for _consumers_ of events. Filesystems cannot by themselves provide any delivery guarantees at all.

Reply:
I think you have a different definition of reliable delivery, by "guarantees for consumers of events", do you mean like force someone to consume when there's none of them?

Reply:
> edit: update docs to say events are best-effort in all cases and everything is fine 👍

in this PR, it only mentions "provides options to write output files reliably", I guess that should be fine.
we can move the discussions of "reliable delivery" to #13652

Reply:
"At least once" [is well-defined](https://doc.akka.io/docs/akka/current/general/message-delivery-reliability.html?language=scala) and is not satisfied by writing output files reliably. If you update docs to say that events are "written to output files reliably" that is also 👍 — as long as it is understood this is not "reliable delivery"

Reply:
> If you update docs to say that events are "written to output files reliably" that is also 👍

it never mentioned reliable delivery in this PR, it's the topic of #13652.